src/server/model_server.py

```diff
@@ -0,0 +1,163 @@
+sys.path.append(util_path)
+sys.path.append(train_path)
+
+from train import get_valid_feature_groups, is_weight_output, ModelOutputType, FeatureGroups, FeatureGroup
+from util.config import getConfig
+from util.loader import parse_filters, is_valid_model, load_json, get_model_weight, get_model_group_path, get_archived_file, METADATA_FILENAME, CHECKPOINT_FOLDERNAME
+
+###############################################
+# model request 
+
+class ModelRequest():
+    def __init__(self, metrics, output_type, source='rapl', node_type=-1, weight=False, trainer_name="", filter=""):
+        self.source = source
+        self.node_type = node_type
+        self.metrics = metrics
+        self.trainer_name = trainer_name
+        self.filter = filter
+        self.output_type = output_type
+        self.weight = weight
+
+###########################################
+
+ERROR_KEY = 'mae'
+ERROR_KEY = getConfig('ERROR_KEY', ERROR_KEY)
+MODEL_SERVER_PORT = 8100
+MODEL_SERVER_PORT = getConfig('MODEL_SERVER_PORT', MODEL_SERVER_PORT)
+MODEL_SERVER_PORT = int(MODEL_SERVER_PORT)
+
+def select_best_model(output_type, valid_groupath, filters, trainer_name="", node_type=-1, weight=False):
+    model_names = [f for f in os.listdir(valid_groupath) if \
+                    f != CHECKPOINT_FOLDERNAME \
+                    and not os.path.isfile(os.path.join(valid_groupath,f)) \
+                    and (trainer_name == "" or trainer_name in f) \
+                    and (node_type == -1 or str(node_type) in f) ]
+    print("Load metadata of trainers:", model_names)
+    best_cadidate = None
+    best_response = None
+    for model_name in model_names:
+        model_savepath = os.path.join(valid_groupath, model_name)
+        metadata = load_json(model_savepath, METADATA_FILENAME)
+        if metadata is None or not is_valid_model(metadata, filters) or ERROR_KEY not in metadata:
+            print("invalid metadata", metadata)
+            continue
+        if weight:
+            response = get_model_weight(valid_groupath, model_name)
+            if response is None:
+                # fail to get weight file
+                print("cannot get model weight", response)
+                continue
+        else:
+            response = get_archived_file(valid_groupath, model_name)
+            if not os.path.exists(response):
+                # archived model file does not exists
+                print("cannot get archived_file", response)
+                continue
+        if best_cadidate is None or best_cadidate[ERROR_KEY] > metadata[ERROR_KEY]:
+            best_cadidate = metadata
+            best_response = response
+    return best_cadidate, best_response
+
+app = Flask(__name__)
+
+@app.route('/model', methods=['POST'])
+def get_model():
+    model_request = request.get_json()
+    req = ModelRequest(**model_request)
+    valid_fgs = get_valid_feature_groups(req.metrics)
+    filters = parse_filters(req.filter)
+    print(req.output_type)
```

Comment:
A few of these print statements may not be necessary and should be removed or logged for future reference.

Reply:
Change from printing to comment. Could we put changing to logging as a new issue and PR applying to other component (such as online trainer) as well?